artiq/coredevice/suservo.py

```diff
@@ -564,3 +564,66 @@ def set_y(self, profile, y):
             raise ValueError("Invalid SUServo y-value!")
         self.set_y_mu(profile, y_mu)
         return y_mu
+
+
+class ProtoRev8(urukul.ProtoRev8):
+    """
+    For use with suservo CPLD to get around polymorphic issues.
+    """
+
+
+class ProtoRev9(urukul.ProtoRev9):
+    """
+    For use with suservo CPLD to get around polymorphic issues.
+    """
+
+
+class CPLD(urukul.CPLD):
+    """
+    Needed to avoid polymorphic issues when using SUServo and regular Urukuls in the same experiment.
+    """
+    rb_len = 8
```

Comment:
This seems related to phase-coherent suservo support and probably shouldn't be in this PR.

Reply:
`nac3` isn't anything close to a drop-in replacement, and in any case this driver is for the legacy compiler only. This PR is just cleaning up after the recent `ProtoRev9` changes, which _were_ merged onto the non-nac3 branch, but broke a common use case (mixing SUServo and non-SUServo Urukuls; cf. various forum discussions).

Reply:
It's not a bug per-se, it's a long-standing limitation. It would be possible to change to "monomorphising" types per instance with some sort of merging of identical shapes, but every Python class corresponding to an ARTIQ Python type is a pretty central part of the current design. The SUServo driver does not take into account this facet of the design (at least not while keeping compatibility with the non-SUServo Urukul driver in the same kernel). This is a real-world issue, as evidenced not just by the Oxford group's, but several other labs' reports here or on the forum, so it obvioussly needs to be fixed.